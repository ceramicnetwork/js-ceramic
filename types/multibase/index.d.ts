declare module 'multibase' {
  export function encode(base: string, payload: Uint8Array): Uint8Array
  export function decode(val: string): Uint8Array
  export function isEncoded(bytes: Uint8Array | string): string
}
