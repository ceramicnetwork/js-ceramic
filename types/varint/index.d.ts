declare module "varint" {
  export const encode: {
    (num: number, array?: Uint8Array, offset?: number): Uint8Array;
    bytes: number;
  }
  export const decode: {
    (buf: Uint8Array, offset?: number): number;
    bytes: number;
  }
}
