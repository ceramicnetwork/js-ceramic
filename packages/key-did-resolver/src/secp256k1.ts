import * as u8a from 'uint8arrays'

/**
 * Constructs the document based on the method key
 */
export function keyToDidDoc(pubKeyBytes: Uint8Array, fingerprint: string): any {
  const did = `did:key:${fingerprint}`
  const keyId = `${did}#${fingerprint}`
  return {
    id: did,
    verificationMethod: [
      {
        id: keyId,
        type: 'Secp256k1VerificationKey2018',
        controller: did,
        publicKeyBase58: u8a.toString(pubKeyBytes, 'base58btc'),
      },
    ],
    authentication: [keyId],
    assertionMethod: [keyId],
    capabilityDelegation: [keyId],
    capabilityInvocation: [keyId],
  }
}
