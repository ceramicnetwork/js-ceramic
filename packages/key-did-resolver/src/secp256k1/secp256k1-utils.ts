import keyto from '@trust/keyto'
import base64url from 'base64url'
import crypto from 'crypto'
import bs58 from 'bs58'
import secp256k1 from 'secp256k1'

import canonicalize from 'canonicalize'

const compressedHexEncodedPublicKeyLength = 66

/** Secp256k1 Private Key  */
// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ISecp256k1PrivateKeyJwk {
    /** key type */
    kty: string;

    /** curve */
    crv: string;

    /** private point */
    d: string;

    /** public point */
    x: string;

    /** public point */
    y: string;

    /** key id */
    kid: string;
}

/** Secp256k1 Public Key  */
// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ISecp256k1PublicKeyJwk {
    /** key type */
    kty: string;

    /** curve */
    crv: string;

    /** public point */
    x: string;

    /** public point */
    y: string;

    /** key id */
    kid: string;
}

/**
 * Example
 * ```js
 * {
 *  kty: 'EC',
 *  crv: 'secp256k1',
 *  d: 'rhYFsBPF9q3-uZThy7B3c4LDF_8wnozFUAEm5LLC4Zw',
 *  x: 'dWCvM4fTdeM0KmloF57zxtBPXTOythHPMm1HCLrdd3A',
 *  y: '36uMVGM7hnw-N6GnjFcihWE3SkrhMLzzLCdPMXPEXlA',
 *  kid: 'JUvpllMEYUZ2joO59UNui_XYDqxVqiFLLAJ8klWuPBw'
 * }
 * ```
 * See [rfc7638](https://tools.ietf.org/html/rfc7638) for more details on Jwk.
 */
export const getKid = (jwk: ISecp256k1PrivateKeyJwk | ISecp256k1PublicKeyJwk): string => {
    const copy = { ...jwk } as any
    delete copy.d
    delete copy.kid
    delete copy.alg
    const digest = crypto
        .createHash('sha256')
        .update(canonicalize(copy))
        .digest()

    return base64url.encode(Buffer.from(digest))
}

/** convert compressed hex encoded private key to jwk */
export const privateKeyJwkFromPrivateKeyHex = (privateKeyHex: string): any => {
    const jwk = {
        ...keyto.from(privateKeyHex, 'blk').toJwk('private'), crv: 'secp256k1',
    }
    const kid = getKid(jwk)
    return {
        ...jwk, kid,
    }
}

/** convert compressed hex encoded public key to jwk */
export const publicKeyJwkFromPublicKeyHex = (publicKeyHex: string): any => {
    let key = publicKeyHex
    if (publicKeyHex.length === compressedHexEncodedPublicKeyLength) {
        const keyBin = secp256k1.publicKeyConvert(Buffer.from(publicKeyHex, 'hex'), false)
        key = Buffer.from(keyBin).toString('hex')
    }
    const jwk = {
        ...keyto.from(key, 'blk').toJwk('public'), crv: 'secp256k1',
    }
    const kid = getKid(jwk)

    return {
        ...jwk, kid,
    }
}

/** convert jwk to hex encoded private key */
export const privateKeyHexFromJwk = async (jwk: ISecp256k1PrivateKeyJwk): Promise<string> => keyto
    .from({
        ...jwk, crv: 'K-256',
    }, 'jwk')
    .toString('blk', 'private')

/** convert jwk to hex encoded public key */
export const publicKeyHexFromJwk = async (jwk: ISecp256k1PublicKeyJwk): Promise<string> => {
    const uncompressedPublicKey = keyto
        .from({
            ...jwk, crv: 'K-256',
        }, 'jwk')
        .toString('blk', 'public')

    const compressed = secp256k1.publicKeyConvert(Buffer.from(uncompressedPublicKey, 'hex'), true)
    return Buffer.from(compressed).toString('hex')
}

/** convert publicKeyHex to base58 */
export const publicKeyBase58FromPublicKeyHex = (publicKeyHex: string) => {
    return bs58.encode(Buffer.from(publicKeyHex, 'hex'))
}

/** convert publicKeyHex to base58 */
export const privateKeyBase58FromPrivateKeyHex = (privateKeyHex: string) => {
    return bs58.encode(Buffer.from(privateKeyHex, 'hex'))
}

export const privateKeyUInt8ArrayFromPrivateKeyBase58 = async (privateKeyBase58: string) => {
    return bs58.decode(privateKeyBase58)
}
