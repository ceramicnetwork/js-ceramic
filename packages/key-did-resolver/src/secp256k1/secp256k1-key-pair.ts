import crypto from 'crypto'

import bs58 from 'bs58'
import secp256k1 from 'secp256k1'
import * as utils from './secp256k1-utils'

/**
 * Encapsulates Secp256k1 key pair
 */
export class Secp256k1KeyPair {
    public id: string
    public type: string
    public controller: string

    public publicKeyBase58: string
    public privateKeyBase58: string

    constructor(options: any = {}) {
        this.type = 'EcdsaSecp256k1VerificationKey2019'

        this.id = options.id
        this.controller = options.controller
        this.privateKeyBase58 = options.privateKeyBase58
        this.publicKeyBase58 = options.publicKeyBase58
    }

    /**
     * Constructs fingerprint from public key
     */
    static fingerprintFromPublicKey({ publicKeyBase58 }: any): string {
        const pubkeyBytes = bs58.decode(publicKeyBase58)
        const buffer = new Uint8Array(2 + pubkeyBytes.length)
        // See https://github.com/multiformats/multicodec/blob/master/table.csv
        // 0xe7 is Secp256k1 public key
        buffer[0] = 0xe7 //
        buffer[1] = 0x01
        buffer.set(pubkeyBytes, 2)
        // prefix with `z` to indicate multi-base base58btc encoding
        return `z${bs58.encode(buffer)}`
    }

    /**
     * Derive instance from options
     */
    static async from(options: any): Promise<Secp256k1KeyPair> {
        let privateKeyBase58 = options.privateKeyBase58
        let publicKeyBase58 = options.publicKeyBase58

        if (options.privateKeyHex) {
            privateKeyBase58 = await utils.privateKeyBase58FromPrivateKeyHex(options.privateKeyHex)
        }

        if (options.publicKeyHex) {
            publicKeyBase58 = await utils.publicKeyBase58FromPublicKeyHex(options.publicKeyHex)
        }

        if (options.privateKeyJwk) {
            privateKeyBase58 = utils.privateKeyBase58FromPrivateKeyHex(await utils.privateKeyHexFromJwk(options.privateKeyJwk))
        }

        if (options.publicKeyJwk) {
            publicKeyBase58 = utils.privateKeyBase58FromPrivateKeyHex(await utils.publicKeyHexFromJwk(options.publicKeyJwk))
        }

        return new Secp256k1KeyPair({
            ...options, privateKeyBase58, publicKeyBase58,
        })
    }

    /**
     * Derive instance from fingerprint
     */
    static fromFingerprint(fingerprint: any): Secp256k1KeyPair {
        // skip leading `z` that indicates base58 encoding
        const buffer = bs58.decode(fingerprint.substr(1))

        // https://github.com/multiformats/multicodec/blob/master/table.csv#L77
        if (buffer[0] === 0xe7 && buffer[1] === 0x01) {
            const publicKeyBase58 = bs58.encode(buffer.slice(2))
            const did = `did:key:${Secp256k1KeyPair.fingerprintFromPublicKey({
                publicKeyBase58,
            })}`
            const keyId = `#${Secp256k1KeyPair.fingerprintFromPublicKey({
                publicKeyBase58,
            })}`
            return new Secp256k1KeyPair({
                id: keyId, controller: did, publicKeyBase58,
            })
        }

        throw new Error(`Unsupported Fingerprint Type: ${fingerprint}`)
    }

    /**
     * Get public key
     */
    get publicKey(): string {
        return this.publicKeyBase58
    }

    /**
     * Get signer functionality
     */
    signer(): any {
        if (!this.privateKeyBase58) {
            return {
                async sign(): Promise<void> {
                    throw new Error('No private key to sign with.')
                },
            }
        }
        const privateKeyBase58 = this.privateKeyBase58
        return {
            async sign({ data }: any): Promise<string> {
                const messageHashUInt8Array = crypto
                    .createHash('sha256')
                    .update(data)
                    .digest()
                const privateKeyUInt8Array = await utils.privateKeyUInt8ArrayFromPrivateKeyBase58(privateKeyBase58)
                const sigObj: any = secp256k1.ecdsaSign(messageHashUInt8Array, privateKeyUInt8Array)

                return sigObj.signature
            },
        }
    }

    /**
     * Get fingerprint from the instance
     */
    fingerprint(): string {
        const { publicKeyBase58 } = this
        return Secp256k1KeyPair.fingerprintFromPublicKey({ publicKeyBase58 })
    }

    /**
     * To JWK
     * @param _private
     */
    toJwk(_private = false): string {
        if (_private) {
            return utils.privateKeyJwkFromPrivateKeyHex(bs58.decode(this.privateKeyBase58).toString('hex'))
        }
        return utils.publicKeyJwkFromPublicKeyHex(bs58.decode(this.publicKeyBase58).toString('hex'))
    }

}
