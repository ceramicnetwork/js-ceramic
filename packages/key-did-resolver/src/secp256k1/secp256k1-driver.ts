import { Secp256k1KeyPair } from "./secp256k1-key-pair"

/**
 * Constructs the document based on the method key
 */
const keyToDidDoc = (fingerprint: string): any => {
    const secp256k1KeyPair = Secp256k1KeyPair.fromFingerprint(fingerprint)
    const did = `did:key:${secp256k1KeyPair.fingerprint()}`
    const keyId = `#${secp256k1KeyPair.fingerprint()}`
    return {
        '@context': ['https://www.w3.org/ns/did/v1', {
            '@base': did,
        },],
        id: did,
        publicKey: [{
            id: keyId,
            type: secp256k1KeyPair.type,
            controller: did,
            publicKeyBase58: secp256k1KeyPair.publicKeyBase58,
        },],
        authentication: [keyId],
        assertionMethod: [keyId],
        capabilityDelegation: [keyId],
        capabilityInvocation: [keyId],
    }
}

/**
 * Resolves single DID
 */
const resolve = async ({ did }: any = {}): Promise<object> => {
    if (!did) {
        throw new TypeError('"did" must be a string.');
    }
    const fingerprint = did
        .split('#')[0]
        .split('did:key:')
        .pop();
    return keyToDidDoc(fingerprint);
}

export {
    keyToDidDoc, resolve
}
