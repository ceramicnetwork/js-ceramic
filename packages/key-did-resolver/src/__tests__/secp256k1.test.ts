import varint from 'varint'
import multibase from 'multibase'
import * as mapper from '../secp256k1'

describe('Secp256k1 mapper', () => {
  it('successfully resolves the document from did', async () => {
    const id = 'zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz'

    const multicodecPubKey = multibase.decode(id)
    varint.decode(multicodecPubKey) // decode is changing param multicodecPubKey as well
    const pubKeyBytes = multicodecPubKey.slice(varint.decode.bytes)
    const doc = await mapper.keyToDidDoc(pubKeyBytes, id)
    expect(doc).toMatchSnapshot()
  })
})
