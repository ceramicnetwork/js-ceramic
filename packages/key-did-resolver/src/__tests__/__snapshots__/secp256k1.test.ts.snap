// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secp256k1 mapper successfully resolves the document from did 1`] = `
Object {
  "assertionMethod": Array [
    "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz#zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
  ],
  "authentication": Array [
    "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz#zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
  ],
  "capabilityDelegation": Array [
    "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz#zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
  ],
  "capabilityInvocation": Array [
    "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz#zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
  ],
  "id": "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
  "verificationMethod": Array [
    Object {
      "controller": "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
      "id": "did:key:zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz#zQ3shbgnTGcgBpXPdBjDur3ATMDWhS7aPs6FRFkWR19Lb9Zwz",
      "publicKeyBase58": "qjdSRUCiVtAjwdYVTxHQXd9FnMPRaYCtKpkchofTw66G",
      "type": "Secp256k1VerificationKey2018",
    },
  ],
}
`;
